### PR TITLE
factor disk attack&add schedule support for disk read&write

### PR DESCRIPTION
**Pull request by Andrewmatilde**
factor all : move Validate process out of attack.
factor disk attack : make Propreocess instead of value validate,check & value preprocess in attack.
Add full support with duration & schedule for disk read&write.

### CODE REVIEW

File: pkg/utils/command.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+			continue
+		}
+
+		if tag == "-" {
+			options = append(options, v.Field(i).String())
+		} else {
+			options = append(options, fmt.Sprintf("%s=%v", tag, v.Field(i).String()))
+		}
+	}
+	return exec.Command(c.Name, options...) //
```
[the remark this answers is not in the document]
Comment (Andrewmatilde) [reply pkg/utils/command.go:42]: ignore

**Comment by Andrewmatilde** [comment]
PTAL, @WangXiangUSTC

**Comment by WangXiangUSTC** [comment]
execute `./bin/chaosd attack disk fill -c 1` or `./bin/chaosd attack disk fill -c 1%` will not run failed with error `fallocate: invalid length value specified`

**Comment by WangXiangUSTC** [comment]
And is it necessary to support the schedule for disk on command mode?

File: pkg/core/disk.go [hunk 2]
```diff
@@ -37,77 +77,195 @@ type DiskOption struct {
...
+		} else {
+			_, err := os.Stat(opt.Path)
+			if err != nil {
 				// check if Path of file is valid when Path is not empty
-				if d.Path != "" {
+				if os.IsNotExist(err) {
 					var b []byte
-					if err := ioutil.WriteFile(d.Path, b, 0644); err != nil {
-						return err
+					if err := ioutil.WriteFile(opt.Path, b, 0600); err != nil {
```
[the remark this answers is not in the document]
Comment (Andrewmatilde) [reply pkg/core/disk.go:194]: Lift suggest me to do that. Change 0644 to 0600 will just reject other user to read this file.

File: pkg/core/disk.go [hunk 3]
```diff
@@ -37,77 +77,195 @@ type DiskOption struct {
...
-						return err
+					if err := os.Remove(opt.Path); err != nil {
+						return "", err
 					}
+				} else {
+					return "", err
 				}
 			} else {
-				return err
+				return "", fmt.Errorf("fill into a existing file")
```
[the remark this answers is not in the document]
Comment (Andrewmatilde) [reply pkg/core/disk.go:204]: OK

File: pkg/core/disk.go [hunk 4]
```diff
@@ -37,77 +77,195 @@ type DiskOption struct {
...
+		if err != nil {
+			log.Error("fail to get disk total size", zap.Error(err))
+			return 0, err
+		}
+		return totalSize * percent / 100, nil
+	}
+	if opt.Action == DiskFillAction {
+		return 0, fmt.Errorf("one of percent and size must not be empty, DiskOption : %v", opt)
 	}
+	return 0, fmt.Errorf("one of size must not be empty, DiskOption : %v", opt)
```
[the remark this answers is not in the document]
Comment (Andrewmatilde) [reply pkg/core/disk.go:270]: OK

File: pkg/core/disk.go [hunk 5]
```diff
@@ -37,77 +77,195 @@ type DiskOption struct {
...
+		if err != nil {
+			log.Error(fmt.Sprintf("fail to get parse size per units , %s", opt.Size), zap.Error(err))
+			return 0, err
+		}
+		return byteSize, nil
+	} else if opt.Percent != "" {
+		opt.Percent = strings.Trim(opt.Percent, " %")
+		percent, err := strconv.ParseUint(opt.Percent, 10, 0)
+		if err != nil {
+			log.Error(fmt.Sprintf(" unexcepted err when parsing disk percent '%s'", opt.Percent), zap.Error(err))
```
[the remark this answers is not in the document]
Comment (Andrewmatilde) [reply pkg/core/disk.go:256]: OK

**Comment by Andrewmatilde** [comment]
> execute `./bin/chaosd attack disk fill -c 1` or `./bin/chaosd attack disk fill -c 1%` will not run failed with error `fallocate: invalid length value specified`

I add `%` support in this version.

**Comment by Andrewmatilde** [comment]
> And is it necessary to support the schedule for disk on command mode?

For schedule , I just put the schedule module writen by shivanshs9 into use.
And I think it is necessary to provide duration control for disk on command mode.

File: pkg/core/disk.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Andrewmatilde) [reply pkg/core/disk.go]: fixed

File: pkg/core/disk.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Andrewmatilde) [reply pkg/core/disk.go]: fixed

**Comment by Andrewmatilde** [comment]
> execute `./bin/chaosd attack disk fill -c 1` or `./bin/chaosd attack disk fill -c 1%` will not run failed with error `fallocate: invalid length value specified`

fixed.

File: pkg/core/disk.go [hunk 6]
```diff
@@ -37,77 +77,195 @@ type DiskOption struct {
...
+	case DiskFillAction:
+		for _, block := range ddBlocks {
+			ddOpts = append(ddOpts, DdOption{
+				ReadPath:  "/dev/zero",
+				WritePath: path,
+				BlockSize: block.BlockSize,
+				Count:     block.Count,
+				Iflag:     "fullblock",
+				Oflag:     "append",
+				Conv:      "notrunc",
```
Comment (WangXiangUSTC) [inline pkg/core/disk.go:148]: how about adding some comment for these different `Conv` value in different action

File: pkg/server/chaosd/disk.go [hunk 7]
```diff
@@ -17,220 +17,111 @@ import (
...
 
 var DiskAttack AttackType = diskAttack{}
 
-const DDWritePayloadCommand = "dd if=/dev/zero of=%s bs=%s count=%s oflag=dsync"
-const DDReadPayloadCommand = "dd if=%s of=/dev/null bs=%s count=%s iflag=dsync,fullblock,nocache"
-
-func (disk diskAttack) Attack(options core.AttackConfig, env Environment) (err error) {
-	attack := options.(*core.DiskOption)
+func (disk diskAttack) Attack(options core.AttackConfig, env Environment) error {
+	if attackConf, ok := options.(*core.DiskAttackConfig); ok {
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/disk.go:37]: how about using `!ok` here and then return error, so we can save some indentation

File: cmd/attack/attack.go [hunk 8]
```diff
@@ -32,3 +36,8 @@ func NewAttackCommand() *cobra.Command {
 
 	return cmd
 }
+
+func SetScheduleFlags(cmd *cobra.Command, conf *core.SchedulerConfig) {
+	cmd.Flags().StringVar(&conf.Duration, "duration", "",
```
[the remark this answers is not in the document]
Comment (Andrewmatilde) [reply cmd/attack/attack.go:41]: I think duration control is an important part of just disk attack . Now user only can use disk attack with space size and speed control is difficult to implement , so I use time control to fill the empty of speed control.

File: pkg/core/disk.go (hunk 6 above)
Comment (Andrewmatilde) [reply pkg/core/disk.go:148]: OK.

File: pkg/server/chaosd/disk.go (hunk 7 above)
Comment (Andrewmatilde) [reply pkg/server/chaosd/disk.go:37]: OK

File: pkg/core/disk.go [hunk 9]
```diff
@@ -37,77 +77,195 @@ type DiskOption struct {
...
-var _ AttackConfig = &DiskOption{}
+func NewDiskOption() *DiskOption {
+	return &DiskOption{
+		CommonAttackConfig: CommonAttackConfig{
+			Kind: DiskAttack,
+		},
+	}
+}
+
+func (opt *DiskOption) PreProcess() (*DiskAttackConfig, error) {
```
[the remark this answers is not in the document]
Comment (Andrewmatilde) [reply pkg/core/disk.go:91]: I think this need more discussion.

File: pkg/server/chaosd/disk.go (hunk 7 above)
Comment (Andrewmatilde) [reply pkg/server/chaosd/disk.go]: fixed.

File: pkg/core/disk.go (hunk 6 above)
Comment (Andrewmatilde) [reply pkg/core/disk.go]: Fixed.

**Comment by Andrewmatilde** [comment]
PTAL, @WangXiangUSTC .

File: pkg/utils/command.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Andrewmatilde) [reply pkg/utils/command.go:42]: ignore

**Review by WangXiangUSTC — APPROVED** [review]
LGTM

**Comment by Andrewmatilde** [comment]
PTAL, @fewdan

**Comment by Andrewmatilde** [comment]
PTAL, @cwen0

**Review by cwen0 — APPROVED** [review]
LGTM

**Pull request merged**